Restore Path cheap copying (refcount/COW PathStream)

smfr opened the pull request:
#### bb3cac815bdf97893a9a5992109a58467a07df06
<pre>
Restore Path cheap copying (refcount/COW PathStream)
<a href="https://bugs.webkit.org/show_bug.cgi?id=258758">https://bugs.webkit.org/show_bug.cgi?id=258758</a>
rdar://111595182

Reviewed by Myles C. Maxfield.

265569@main undid the &quot;copy on write&quot; behavior of class; assigning a Path would copy
the underlying PathStream.

To fix this, store the Vector&lt;PathSegment&gt; in an internal, thread-safe ref-counted SegmentsData
struct, and have PathStream hold a DataRef to this struct.

All functions that mutate the path segments call `access()` on the DataRef to trigger a copy.

* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::create):
(WebCore::PathStream::PathStream):
(WebCore::PathStream::clone const):
(WebCore::PathStream::operator== const):
(WebCore::PathStream::lastIfMoveTo const):
(WebCore::PathStream::moveTo):
(WebCore::PathStream::addLineTo):
(WebCore::PathStream::addQuadCurveTo):
(WebCore::PathStream::addBezierCurveTo):
(WebCore::PathStream::addArcTo):
(WebCore::PathStream::addArc):
(WebCore::PathStream::addRect):
(WebCore::PathStream::addEllipse):
(WebCore::PathStream::addEllipseInRect):
(WebCore::PathStream::addRoundedRect):
(WebCore::PathStream::closeSubpath):
(WebCore::PathStream::segments const):
(WebCore::PathStream::applySegments const):
(WebCore::PathStream::applyElements const):
(WebCore::PathStream::singleSegment const):
(WebCore::PathStream::isClosed const):
(WebCore::PathStream::currentPoint const):
(WebCore::PathStream::fastBoundingRect const):
(WebCore::PathStream::boundingRect const):
* Source/WebCore/platform/graphics/PathStream.h:

Canonical link: <a href="https://commits.webkit.org/265810@main">https://commits.webkit.org/265810@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eabf3c0529aa56f4828c3da3c554b1d06650f4df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11967 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12313 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12616 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13713 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/11550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11986 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12327 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14249 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12131 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12944 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10125 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14125 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10236 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10861 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17987 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11314 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11020 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14187 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11502 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9460 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10716 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2922 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15041 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11360 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/platform/graphics/PathStream.cpp:
```diff
@@ -134,56 +150,57 @@ void PathStream::addArc(const FloatPoint& point, float radius, float startAngle,
 void PathStream::addRoundedRect(const FloatRoundedRect& roundedRect, PathRoundedRect::Strategy strategy)
 {
-    m_segments.append(PathRoundedRect { roundedRect, strategy });
+    m_segmentsData.access().segments.append(PathRoundedRect { roundedRect, strategy });
```
Is there no way for the compiler to enforce that we call `access()` in all the right places?

smfr replied:
I can try making `segments` private and provide only a `const` accessor.

litherum commented on Source/WebCore/platform/graphics/PathStream.h:
```diff
@@ -70,6 +72,44 @@ class PathStream final : public PathImpl {
+            return segments == other.segments;
+        }
+
+        Vector<PathSegment> segments;
```
@heycam makes a good point that this may want some inline capacity.

smfr replied:
But making empty paths is probably common.

heycam replied:
True, but we already were spending one path component's worth of storage before Said's patch. (Well, maybe a bit less than the new PathSegment type, but still a few words.)

nikolaszimmermann commented on Source/WebCore/platform/graphics/PathStream.h:
```diff
@@ -83,12 +123,14 @@ class PathStream final : public PathImpl {
     std::optional<PathDataBezierCurve> singleBezierCurve() const final;
 
-    bool isEmpty() const final { return m_segments.isEmpty(); }
+    bool isEmpty() const final { return segments().isEmpty(); }
```
Calling segments(), calls DataRef::access() - does that trigger a copy? From the commit message I would expect only non-const functions to be able to call segments()? I might confuse something...

smfr replied:
There's a `const` version of `segments()` higher up in the header, but it's exported so not inline. I'll change this one to access the data member directly.

nikolaszimmermann replied:
Ah it wouldn't even have compiled without the presence of a non-const accessor, missed that. Thanks Simon!

smfr committed:
Restore Path cheap copying (refcount/COW PathStream)
https://bugs.webkit.org/show_bug.cgi?id=258758
rdar://111595182

Reviewed by Myles C. Maxfield.

265569@main undid the "copy on write" behavior of class; assigning a Path would copy
the underlying PathStream.

To fix this, store the Vector<PathSegment> in an internal, thread-safe ref-counted SegmentsData
struct, and have PathStream hold a DataRef to this struct.

All functions that mutate the path segments call `access()` on the DataRef to trigger a copy.

* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::create):
(WebCore::PathStream::PathStream):
(WebCore::PathStream::clone const):
(WebCore::PathStream::operator== const):
(WebCore::PathStream::lastIfMoveTo const):
(WebCore::PathStream::moveTo):
(WebCore::PathStream::addLineTo):
(WebCore::PathStream::addQuadCurveTo):
(WebCore::PathStream::addBezierCurveTo):
(WebCore::PathStream::addArcTo):
(WebCore::PathStream::addArc):
(WebCore::PathStream::addRect):
(WebCore::PathStream::addEllipse):
(WebCore::PathStream::addEllipseInRect):
(WebCore::PathStream::addRoundedRect):
(WebCore::PathStream::closeSubpath):
(WebCore::PathStream::segments const):
(WebCore::PathStream::applySegments const):
(WebCore::PathStream::applyElements const):
(WebCore::PathStream::singleSegment const):
(WebCore::PathStream::isClosed const):
(WebCore::PathStream::currentPoint const):
(WebCore::PathStream::fastBoundingRect const):
(WebCore::PathStream::boundingRect const):
* Source/WebCore/platform/graphics/PathStream.h:

Canonical link: https://commits.webkit.org/265810@main
